### DISCUSSION TITLE
Is there an MAUI Preferences equivalent for avalonia.

### DISCUSSION THREAD

**Discussion by DirkKramer** [AvaloniaUI/Avalonia#9915] [Q&A]
I am creating a multiplatform application for (Windows, Android etc..), in this application i want to store some user specific strings.
I tried using System.ConfigurationManager without success.

 When I tried MAUI there was an possibility to store user specific parameters using the Preferences class.  Does Avalonia have a way to do this?

Thanks in advance!

**Comment by workgroupengineering** [comment]
Using [Xamarin.Essentials](https://github.com/xamarin/Essentials#xamarinessentials)

**Reply by pzs7602** [reply]
If you want to store not so much data, you can use [JsonSerializer Class](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer?view=net-7.0). This is .NET API so is suitable for cross-platform application.

**Comment by DirkKramer** [comment]
I tried using both Xamarin and Maui Essentials.  Both dont work on a cross platform avalonia solution.

I added this to my project file:     ` <UseMauiEssentials>true</UseMauiEssentials>` without succes.


This is the error I get:
![image](https://user-images.githubusercontent.com/46933603/211601234-16187e9a-2ccc-49f9-ae67-c2b9de996226.png)

**Reply by maxkatz6** [reply]
Maui Essentials has implementations only for mobile platforms. There is no implementation for desktop platform if you use it there.

Other than that, you can reference platform specific packages in your project. For instance:
https://www.nuget.org/packages/Microsoft.Maui.Essentials.Runtime.android

I would expect UseMauiEssentials to do it automatically, depending on your target frameworks, but not sure, maybe it doesn't work in all cases. And maybe you need to reference nuget packages per platform independently.

**Reply by maxkatz6** [reply]
For non mobiles OS, as Maui Essentials don't support them, you probably will have to use another implementation with json files or some abstraction like Microsoft.Extensions.Configuration. You can research these options and choose.
You can have feature interface and multiple implementations for it.

**Comment by DrWenz** [comment]
Is there any guide to use Xamarin.Essentials or Maui.Essentials in AvaloniaUI?
I always get NotImplementedInReferenceAssemblyException :(

**Reply by maxkatz6** [reply]
With .NET 8 you can reference https://www.nuget.org/packages/Microsoft.Maui.Essentials/8.0.0-preview.3.8149#versions-body-tab

Before .NET 8 (and after as well) `<UseMauiEssentials>true</UseMauiEssentials>` should be enough, but I guess there is a bug in MAUI or something.

**Reply by JGKle** [reply]
Is there a complete example of how to initialize this in Avalonia? Their official docs (https://learn.microsoft.com/en-us/dotnet/maui/migration/native-essentials?view=net-maui-8.0&tabs=ios#initialize-the-platform) say to override FinishedLaunching, but AvaloniaAppDelegate doesn't let you do that because it's not virtual. Can't find any documentation or examples that actually works with Avalonia.

**Reply by jp2masa** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/14234#discussioncomment-9076845

**Reply by thevortexcloud** [reply]
The official Avalonia MAUI repo has instructions.

https://github.com/AvaloniaUI/AvaloniaMauiHybrid

**Reply by JGKle** [reply]
Thanks!

**Comment by Mrxx99** [comment]
https://github.com/sandreas/Avalonia.Preferences
